data/txHandler.go

```diff
@@ -623,32 +630,102 @@ func (handler *TxHandler) incomingMsgDupCheck(data []byte) (*crypto.Digest, bool
 	return msgKey, false
 }
 
+// erlClientMapper handles erlIPClient creation from erlClient
+// in order to map multiple clients to a single IP address.
+// Then that meta erlIPClient is supposed to be supplied to ERL
+type erlClientMapper struct {
+	m          deadlock.RWMutex
+	mapping    map[string]*erlIPClient
+	maxClients int
+}
+
+// getClient returns erlIPClient for a given sender
+func (mp *erlClientMapper) getClient(sender network.DisconnectableAddressablePeer) util.ErlClient {
+	addr := string(sender.RoutingAddr())
+	ec := sender.(util.ErlClient)
+
+	mp.m.Lock()
+	ipClient, has := mp.mapping[addr]
+	if !has {
+		ipClient = &erlIPClient{
+			clients: make(map[util.ErlClient]struct{}, mp.maxClients),
+		}
+		mp.mapping[addr] = ipClient
+	}
+	mp.m.Unlock()
+
+	ipClient.register(ec)
+	return ipClient
+}
+
+type erlIPClient struct {
+	util.ErlClient
+	m       deadlock.RWMutex
+	clients map[util.ErlClient]struct{}
+	closer  atomic.Value
+}
+
+func (eic *erlIPClient) OnClose(f func()) {
+	eic.closer.Store(f)
+}
+
+// register registers a new client to the erlIPClient
+// by adding a helper closer function to track connection closures
+func (eic *erlIPClient) register(ec util.ErlClient) {
+	eic.m.Lock()
+	_, has := eic.clients[ec]
+	defer eic.m.Unlock()
+	if has {
+		// this peer is known => noop
+		return
+	}
+	eic.clients[ec] = struct{}{}
+
+	ec.OnClose(func() {
+		eic.connClosed(ec)
+	})
+}
+
+// connClosed is called when a connection is closed so that
+// erlIPClient removes the client from its list of clients
+// and calls the closer function if there are no more clients
+func (eic *erlIPClient) connClosed(ec util.ErlClient) {
+	eic.m.Lock()
+	delete(eic.clients, ec)
+	empty := len(eic.clients) == 0
+	eic.m.Unlock()
+	// if no elements left, call the closer
+	// use atomic Swap in order to retrieve the closer and call it once
+	if empty {
+		// atomic.Value.Swap does not like nil values so use a typed nil
+		// and cast it to func() in order to compare with nil
+		if closer := eic.closer.Swap((func())(nil)).(func()); closer != nil {
+			closer()
+		}
+	}
+}
```

Comment:
Why do you use both a mutex and an atomic?  Could you protect the `eic.clients` _and_ the `eic.closer` with the `eic.m` mutex, so you don't have to do the special handling/casting of the atomic?

Reply:
to reduce the mutex congestion: the close function could be relatively expensive and require own locks.

Reply:
agreed on this case scenario, removed atomic, moved inside a locker scope.